# Executive Summary

## Project Overview
This project develops a Machine Learning solution to predict individual medical insurance charges using customer demographic and lifestyle information. Multiple boosting regression algorithms are trained and evaluated to identify the model that provides the most accurate predictions.

## Objective
- Predict medical insurance charges using historical customer data.
- Train and evaluate multiple regression models.
- Compare model performance using regression metrics.
- Select the best-performing boosting model for deployment.

## Approach
1. Data Loading
2. Exploratory Data Analysis (EDA)
3. Data Preprocessing
4. Feature Engineering and Encoding
5. Train-Test Split
6. Model Training
7. Model Evaluation and Comparison
8. Best Model Selection
9. Conclusion

## Outcome
The project compares multiple regression models and selects the best-performing model based on evaluation metrics, demonstrating an end-to-end machine learning workflow suitable for real-world business applications.

# Dataset Description

| Feature | Type | Description |
|---|---|---|
| age | Numerical | Age of the policyholder |
| sex | Categorical | Gender |
| bmi | Numerical | Body Mass Index |
| children | Numerical | Number of dependents |
| smoker | Categorical | Smoking status |
| region | Categorical | Residential region |
| charges | Target | Annual medical insurance charges |


# 📚 Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# 📂 Load Dataset

In [2]:
# Load the input data to pandas dataframe.
df = pd.read_csv("../data/insurance_pre.csv")
df.head()

,age,sex,bmi,children,smoker,charges
0,19,female,27.900,0,yes,16884.92400
1,18,male,33.770,1,no,1725.55230
2,28,male,33.000,3,no,4449.46200
3,33,male,22.705,0,no,21984.47061
4,32,male,28.880,0,no,3866.85520


# 🔍 Exploratory Data Analysis (EDA)

In [3]:
# Get the number of rows and columns in the dataset.
df.shape

(1338, 6)

- There are 1338 rows and 6 coluns in the dataset.

In [4]:
# Get the information about the datatype of each column in the dataset.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 62.8+ KB


- There are fours numberical columns (two columns with interger data and two columns with floating data)
- Two columns with categorical data ("sex" and "smoker" are nominal type of data).

In [5]:
# Check the missing values in the columns of the dataset.
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
charges     0
dtype: int64

- There is no empty value in the entire dataset.

In [6]:
# Check if there is any duplicate in the dataset.
df.duplicated().sum()

np.int64(1)

- There is one row with duplicate value in the dataset.

In [7]:
# Display the duplicate rows in the dataset.
df[df.duplicated(keep=False)]

,age,sex,bmi,children,smoker,charges
195,19,male,30.59,0,no,1639.5631
581,19,male,30.59,0,no,1639.5631


In [8]:
# Remove the duplicate row(s) in the dataset.
df = df.drop_duplicates()

# 🧹 Data Preprocessing

In [9]:
# Convert the nominal data in the columns such as "sex" and "smoker".
df = pd.get_dummies(df, columns=['sex', 'smoker'], dtype=int, drop_first=True)

In [10]:
# Print first five rows to show the nominal values are converted to integers.
df.head()

,age,bmi,children,charges,sex_male,smoker_yes
0,19,27.900,0,16884.92400,0,1
1,18,33.770,1,1725.55230,1,0
2,28,33.000,3,4449.46200,1,0
3,33,22.705,0,21984.47061,1,0
4,32,28.880,0,3866.85520,1,0


# # 🎯 Define Features and Target

In [11]:
X = df.drop("charges", axis=1)
y = df["charges"]

# ✂️ Train-Test Split

In [12]:
# Import train_test_split function from sklearn library.
from sklearn.model_selection import train_test_split

# Split the dataset into 70% for training and 30% for testing.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=42)

# Regression Models

## 1. Ada Boost Regression model

In [13]:
from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score

# Base estimator (weak learner)
base_model = DecisionTreeRegressor(
    max_depth=3,
    random_state=42
)

# AdaBoost Regressor
ada_boost_model = AdaBoostRegressor(
    estimator=base_model,
    n_estimators=100,
    learning_rate=0.1,
    random_state=42
)

# Train
ada_boost_model.fit(X_train, y_train)

# Predict
y_pred = ada_boost_model.predict(X_test)

# Evaluate
print("R² Score:", r2_score(y_test, y_pred))

R² Score: 0.8447927324712745


## 2. XG Boost model

In [14]:
from xgboost import XGBRegressor
from sklearn.metrics import r2_score

# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Create model
xg_boost_model = XGBRegressor(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)

# Train
xg_boost_model.fit(X_train, y_train)

# Predict
y_pred = xg_boost_model.predict(X_test)

# Evaluate
print("R² Score:", r2_score(y_test, y_pred))

R² Score: 0.9001482603478119


## 3. LG Boost model

In [15]:
from lightgbm import LGBMRegressor

lg_boost_model = LGBMRegressor(
    num_leaves=31,
    max_depth=-1,          # No depth limit
    min_child_samples=10,  # Lower than the default 20
    learning_rate=0.05,
    n_estimators=300,
    random_state=42,
    verbose=-1             # Hides informational logs and warnings
)

# Train
lg_boost_model.fit(X_train, y_train)

# Predict
y_pred = lg_boost_model.predict(X_test)

# Evaluate
print("R² Score:", r2_score(y_test, y_pred))

R² Score: 0.8777041052413209


# Save The Best Model

In [16]:
## Save the model to pickle library.
import pickle

# Save the best trained model to a pickle file
with open("../models/final_model.sav", "wb") as file:
    pickle.dump(xg_boost_model, file)

print("Model saved successfully!")

Model saved successfully!


# Model Comparison

| Model | R² Score | Selected |
|---|---:|:---:|
| Ada Boost | 0.84479 | |
| XG Boost | **0.9001** |✅|
| LG Boost | 0.8777 | |



# Final Conclusion with Business Insights

## Conclusion
Among the evaluated regression models, **XG Boost Regressor** achieved the highest R² score (0.9001), making it the best-performing model for predicting medical insurance charges in this project.

## Business Insights
- Supports more accurate premium estimation.
- Helps identify high-cost customers.
- Enables data-driven pricing decisions.
- Reduces manual effort in insurance cost estimation.
